pkg/skaffold/test/custom/custom.go

```diff
@@ -0,0 +1,161 @@
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/schema/latest"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/util"
+)
+
+type Runner struct {
+	customTest     latest.CustomTest
+	testWorkingDir string
+	localDaemon    docker.LocalDaemon
+}
+
+// New creates a new custom.Runner.
+func New(cfg docker.Config, wd string, ct latest.CustomTest) (*Runner, error) {
+	localDaemon, err := docker.NewAPIClient(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Runner{
+		customTest:     ct,
+		testWorkingDir: wd,
+		localDaemon:    localDaemon,
+	}, nil
+}
+
+// Test is the entrypoint for running custom tests
+func (ct *Runner) Test(ctx context.Context, out io.Writer, _ []build.Artifact) error {
+	if err := ct.runCustomCommand(ctx, out); err != nil {
+		return fmt.Errorf("running custom test command: %w", err)
+	}
+
+	return nil
+}
+
+func (ct *Runner) runCustomCommand(ctx context.Context, out io.Writer) error {
+	test := ct.customTest
+
+	// Expand command
+	command, err := util.ExpandEnvTemplate(test.Command, nil)
+	if err != nil {
+		return fmt.Errorf("unable to parse test command %q: %w", test.Command, err)
+	}
+
+	logrus.Infof("Running custom test command %s", command)
+	color.Default.Fprintln(out, "Running custom test command: ", command)
+	color.Default.Fprintln(out, "Custom test timeout is set to: ", test.TimeoutSeconds)
+
+	if test.TimeoutSeconds > 0 {
+		newCtx, cancel := context.WithTimeout(ctx, (time.Duration(test.TimeoutSeconds))*(time.Second))
+		defer cancel()
+		ctx = newCtx
+	}
+
+	var cmd *exec.Cmd
+	// We evaluate the command with a shell so that it can contain
+	// env variables.
+	if runtime.GOOS == "windows" {
+		cmd = exec.CommandContext(ctx, "cmd.exe", "/C", command)
+	} else {
+		cmd = exec.CommandContext(ctx, "sh", "-c", command)
+	}
+
+	cmd.Stdout = out
+	cmd.Stderr = out
+	cmd.Env = ct.env()
+
+	if err := cmd.Start(); err != nil {
+		return fmt.Errorf("error running cmd: %w", err)
+	}
+
+	if err := cmd.Wait(); err != nil {
+		if e, ok := err.(*exec.ExitError); ok {
+			// If the process exited by itself, just return the error
+			if e.Exited() {
+				return e
+			}
+			// If the context is done, it has been killed by the exec.Command
+			select {
+			case <-ctx.Done():
+				if ctx.Err() == context.DeadlineExceeded {
+					fmt.Println("Command timed out")
+				}
+				return ctx.Err()
+			default:
+				return e
+			}
+		}
+		return err
+	}
+	return nil
+}
+
+// env returns a merged environment of the current process environment and any extra environment.
+func (ct *Runner) env() []string {
+	extraEnv := ct.localDaemon.ExtraEnv()
+	if extraEnv == nil {
+		return nil
+	}
+
+	parentEnv := os.Environ()
+	mergedEnv := make([]string, len(parentEnv), len(parentEnv)+len(extraEnv))
+	copy(mergedEnv, parentEnv)
+	return append(mergedEnv, extraEnv...)
+}
+
+// TestDependencies returns dependencies listed for a custom test
+func (ct *Runner) TestDependencies() ([]string, error) {
```

Comment:
To disambiguate this from a normal Test* function, maybe consider naming this `ListTestDependencies()`?

Reply:
This is an interface method that's being called from New() in dev loop.